libs/gretty/src/main/groovy/org/akhikhl/gretty/ProjectUtils.groovy

```diff
@@ -84,20 +84,37 @@ final class ProjectUtils {
   }
 
   private static void addProjectClassPathWithDependencies(Set<URL> urls, Project proj, String dependencyConfigName) {
-    if(proj.sourceSets.main.output.files.any { new File(it, 'META-INF/web-fragment.xml').exists() })
-      urls.add proj.jar.archivePath.toURI().toURL()
-    else
-      urls.addAll proj.sourceSets.main.output.files.collect { it.toURI().toURL() }
     def dependencyConfig = proj.configurations.findByName(dependencyConfigName)
     if(dependencyConfig) {
-      def files = dependencyConfig.fileCollection { !(it instanceof ProjectDependency) }
+      Set<File> dependencyProjectArchives = new LinkedHashSet<>()
+
+      def collectProjectDependencies
```

Comment:
There is an awful lot going on within this method. Actually, I would review this by leaving a couple of remarks, and leave the resolution to you. This time it did not work for me. Because this method is so complicated (I had difficulties to see how the behavior changed), I had to dissect it myself and debug multiple times. After that, I had come up with the following refactor:

```groovy
private static void addProjectClassPathWithDependencies(Set<URL> urls, Project project, String dependencyConfigName) {
  Set<File> files = project.configurations.findByName(dependencyConfigName)?.resolve() ?: []

  def (outputs, archives) = collectOutputsAndArchivesWithDependencies(project, dependencyConfigName)
  files += outputs
  files -= archives

  def grettyProvidedCompileConfig = project.configurations.findByName('grettyProvidedCompile').collect()
  files -= grettyProvidedCompileConfig

  urls.addAll(files.collect { it.toURI().toURL() })
}

private static collectOutputsAndArchivesWithDependencies(Project rootProject, String dependencyConfigName) {
  Set<File> outputs = []
  Set<File> archives = []

  walkProjectDependencies(rootProject, dependencyConfigName) { project ->
    def outputFiles = project.sourceSets.main.output.files
    if(outputFiles.any { new File(it, 'META-INF/web-fragment.xml').exists() }) {
      outputs.add(project.jar.archivePath)
    } else {
      outputs.addAll(outputFiles)
      archives.add(project.jar.archivePath)
    }
  }

  return new Tuple2(outputs, archives)
}

private static void walkProjectDependencies(Project project, String dependencyConfigName, Closure projectClosure) {
  def dependencyConfig = project.configurations.findByName(dependencyConfigName)
  if (!dependencyConfig) return
  dependencyConfig.allDependencies.withType(ProjectDependency).each {
    projectClosure(it.dependencyProject)
    walkProjectDependencies(it.dependencyProject, dependencyConfigName, projectClosure)
  }
}
```

Let's flip the review. Please tell me if you think the code is more readable (understandable), if the piece still covers all assumptions you made when coding your proposal (correctness), and what you would like to change about it. In the meantime I'll let the test suite run with my proposed changes.